### PR TITLE
Fast shared mutex integration

### PR DESCRIPTION
**Pull request by serxa**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Switched to faster shared (RW) mutex implementation. Performance may be improved in queries with a lot of thread synchronization or for data structures experiencing heavy contention.

### CODE REVIEW

**Comment by serxa** [comment]
Should be merged after https://github.com/ClickHouse/ClickHouse/pull/44924

File: src/Dictionaries/CacheDictionary.h [hunk 1]
```diff
@@ -206,7 +206,7 @@ class CacheDictionary final : public IDictionary
     /// This lock is used for the inner cache state update function lock it for
     /// write, when it need to update cache state all other functions just
     /// readers. Surprisingly this lock is also used for last_exception pointer.
-    mutable std::shared_mutex rw_lock;
+    mutable DB::FastSharedMutex rw_lock;
```
Comment (rschu1ze) [inline src/Dictionaries/CacheDictionary.h:209]: (could omit `DB::`)

File: src/Dictionaries/CacheDictionaryUpdateQueue.h [hunk 2]
```diff
@@ -2,12 +2,12 @@
 
 #include <atomic>
 #include <mutex>
-#include <shared_mutex>
 #include <utility>
 #include <vector>
 #include <functional>
 
 #include <Common/ThreadPool.h>
+#include <Common/Threading.h>
```
Comment (rschu1ze) [inline src/Dictionaries/CacheDictionaryUpdateQueue.h:10]: Forgot to replace std mutex by the new one in this file?

**Comment by rschu1ze** [comment]
Looks overall good to me. Will check the lock code itself (#44924) separately.

**Review by rschu1ze — APPROVED** [review]

**Comment by serxa** [comment]
`01193_metadata_loading` stateless test failed in release looks a little bit suspicious, but I think it is unrelated, let's rerun the tests. 
```
2023-01-13 08:18:09 [e73d747685c1] 2023.01.13 08:17:51.056919 [ 119663 ] {acc31595-6ae6-4f89-b33b-078b4242d170}  executeQuery: Code: 81. DB::Exception: Database test_01193_7751 is currently dropped or renamed. (UNKNOWN_DATABASE) (version 22.13.1.2082) (from [::1]:58260) (comment: 01193_metadata_loading.sh) (in query: CREATE TABLE test_01193_7751.table_2_896 (i UInt64, d Date, s String, n Nested(i UInt8, f Float32)) ENGINE=File(CSV);), Stack trace (when copying this message, always include the lines below):
2023-01-13 08:18:09 
2023-01-13 08:18:09 0. ./build_docker/../src/Common/Exception.cpp:77: DB::Exception::Exception(DB::Exception::MessageMasked const&, int, bool) @ 0xf486d5a in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 1. DB::Exception::Exception, std::__1::allocator> const&>(int, fmt::v8::basic_format_string, std::__1::allocator> const&>::type>, std::__1::basic_string, std::__1::allocator> const&) @ 0x8e85a20 in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 2. ./build_docker/../src/Interpreters/DatabaseCatalog.cpp:1300: DB::DDLGuard::DDLGuard(std::__1::map, std::__1::allocator>, DB::DDLGuard::Entry, std::__1::less, std::__1::allocator>>, std::__1::allocator, std::__1::allocator> const, DB::DDLGuard::Entry>>>&, DB::SharedMutex&, std::__1::unique_lock, std::__1::basic_string, std::__1::allocator> const&, std::__1::basic_string, std::__1::allocator> const&) @ 0x142626bd in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 3. ./build_docker/../contrib/llvm-project/libcxx/include/__mutex_base:143: DB::DatabaseCatalog::getDDLGuard(std::__1::basic_string, std::__1::allocator> const&, std::__1::basic_string, std::__1::allocator> const&) @ 0x142563bc in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 4. ./build_docker/../contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:302: DB::InterpreterCreateQuery::doCreateTable(DB::ASTCreateQuery&, DB::InterpreterCreateQuery::TableProperties const&, std::__1::unique_ptr>&) @ 0x149a226a in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 5. ./build_docker/../src/Interpreters/InterpreterCreateQuery.cpp:0: DB::InterpreterCreateQuery::createTable(DB::ASTCreateQuery&) @ 0x1499d7f2 in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 6. ./build_docker/../src/Interpreters/InterpreterCreateQuery.cpp:1635: DB::InterpreterCreateQuery::execute() @ 0x149a76cd in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 7. ./build_docker/../src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::__1::shared_ptr, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x14f3e493 in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 8. ./build_docker/../src/Interpreters/executeQuery.cpp:1108: DB::executeQuery(std::__1::basic_string, std::__1::allocator> const&, std::__1::shared_ptr, bool, DB::QueryProcessingStage::Enum) @ 0x14f3b46d in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 9. ./build_docker/../src/Server/TCPHandler.cpp:0: DB::TCPHandler::runImpl() @ 0x15c3468f in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 10. ./build_docker/../src/Server/TCPHandler.cpp:1937: DB::TCPHandler::run() @ 0x15c48559 in /usr/lib/debug/usr/bin/clickhouse.debug
2023-01-13 08:18:09 11. ./build_docker/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConne
```

**Comment by serxa** [comment]
I was wrong, failures are related. Concurrent `try_lock_shared()` could fail with the current implementation. I'll fix it.

**Comment by serxa** [comment]
Failed tests are now all unrelated:
Integration Tests (asan) - test_azure_blob_storage_zero_copy_replication/test.py::test_zero_copy_replication 
Integration Tests (asan) - test_cluster_copier/test.py::test_copy_with_recovering[True]
Integration Tests (release) - test_merge_tree_s3_restore/test.py::test_restore_to_detached[False-True]

**Pull request merged**